ethers-etherscan/src/lib.rs

```diff
@@ -0,0 +1,454 @@
+pub struct ContractMetadata {
+    #[serde(flatten)]
+    pub items: Vec<Metadata>,
+}
+
+impl IntoIterator for ContractMetadata {
+    type Item = Metadata;
+    type IntoIter = std::vec::IntoIter<Metadata>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.items.into_iter()
+    }
+}
+
+impl ContractMetadata {
+    /// All ABI from all contracts in the source file
+    pub fn abis(&self) -> anyhow::Result<Vec<Abi>> {
+        let mut abis = Vec::with_capacity(self.items.len());
+        for item in &self.items {
+            abis.push(serde_json::from_str(&item.abi)?);
+        }
+        Ok(abis)
+    }
+
+    /// Combined source code of all contracts
+    pub fn source_code(&self) -> String {
+        self.items
+            .iter()
+            .map(|c| c.source_code.as_str())
+            .collect::<Vec<_>>()
+            .join("\n")
+    }
+}
+
+/// Etherscan contract metadata
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Metadata {
+    #[serde(rename = "SourceCode")]
+    pub source_code: String,
+    #[serde(rename = "ABI")]
+    pub abi: String,
+    #[serde(rename = "ContractName")]
+    pub contract_name: String,
+    #[serde(rename = "CompilerVersion")]
+    pub compiler_version: String,
+    #[serde(rename = "OptimizationUsed")]
+    pub optimization_used: String,
+    #[serde(rename = "Runs")]
+    pub runs: String,
+    #[serde(rename = "ConstructorArguments")]
+    pub constructor_arguments: String,
+    #[serde(rename = "EVMVersion")]
+    pub evm_version: String,
+    #[serde(rename = "Library")]
+    pub library: String,
+    #[serde(rename = "LicenseType")]
+    pub license_type: String,
+    #[serde(rename = "Proxy")]
+    pub proxy: String,
+    #[serde(rename = "Implementation")]
+    pub implementation: String,
+    #[serde(rename = "SwarmSource")]
+    pub swarm_source: String,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn api_key() -> String {
+        std::env::var("ETHERSCAN_API_KEY").expect("ETHERSCAN_API_KEY not found")
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn can_fetch_contract_abi() {
+        let api = api_key();
+        let client = Client::new("mainnet", api).unwrap();
+
+        let _abi = client
+            .contract_abi(
+                "0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413"
+                    .parse()
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn can_fetch_contract_source_code() {
+        let api = api_key();
+        let client = Client::new("mainnet", api).unwrap();
+
+        let _meta = client
+            .contract_source_code(
+                "0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413"
+                    .parse()
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn can_verify_contract() {
+        // TODO this needs further investigation
```

Comment:
Is this working now or does it still require more investigation?

Reply:
still get this response _instantly_, so I think I messed up the payload somehow, but tried various, alternatives like numbers as strings, etc.., will need to take a closer look at dapptools itself or any other etherscan client
```
    Response {
        status: "0",
        message: "NOTOK",
        result: "Error!",
    },
```